File: app/src/main/java/com/futsch1/medtimer/database/JSONMedicineBackup.java
package com.futsch1.medtimer.database;

import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;

import java.time.Instant;
import java.util.List;

public class JSONMedicineBackup extends JSONBackup<FullMedicine> {

    public JSONMedicineBackup() {
        super(FullMedicine.class);
    }

    @Override
    public JsonElement createBackup(int databaseVersion, List<FullMedicine> list) {
        // Fix the medicines where the instructions are null
        for (FullMedicine FullMedicine : list) {
            FullMedicine.reminders.stream().filter(reminder -> reminder.instructions == null).forEach(reminder -> reminder.instructions = "");
        }
        return super.createBackup(databaseVersion, list);
    }

    @Override
    protected GsonBuilder registerTypeAdapters(GsonBuilder builder) {
        return builder
                .registerTypeAdapter(Medicine.class, new FullDeserialize<Medicine>())
                .registerTypeAdapter(Tag.class, new FullDeserialize<Tag>())
                .registerTypeAdapter(Reminder.class, new FullDeserialize<Reminder>());
    }

    protected boolean isInvalid(FullMedicine fullMedicine) {
        return fullMedicine == null || fullMedicine.medicine == null || fullMedicine.reminders == null;
    }

    public void applyBackup(List<FullMedicine> listOfFullMedicine, MedicineRepository medicineRepository) {
        medicineRepository.deleteReminders();
        medicineRepository.deleteMedicines();

        for (FullMedicine FullMedicine : listOfFullMedicine) {
            long medicineId = medicineRepository.insertMedicine(FullMedicine.medicine);
            processReminders(medicineRepository, FullMedicine, (int) medicineId);
            processTags(medicineRepository, FullMedicine, (int) medicineId);
        }
    }

    private static void processReminders(MedicineRepository medicineRepository, FullMedicine fullMedicine, int medicineId) {
        for (Reminder reminder : fullMedicine.reminders) {
            if (reminder != null) {
                reminder.medicineRelId = medicineId;
                reminder.createdTimestamp = Instant.now().toEpochMilli() / 1000;
                medicineRepository.insertReminder(reminder);
            }
        }
    }

    private void processTags(MedicineRepository medicineRepository, FullMedicine fullMedicine, int medicineId) {
        for (Tag tag : fullMedicine.tags) {
            Tag existingTag = medicineRepository.getTagByName(tag.name);
            if (existingTag != null) {
                medicineRepository.insertMedicineToTag(medicineId, existingTag.tagId);
                continue;
            }
            int tagId = (int) medicineRepository.insertTag(tag);
            medicineRepository.insertMedicineToTag(medicineId, tagId);
        }
    }
}

File: app/src/main/java/com/futsch1/medtimer/database/JSONMedicineBackup.kt
package com.futsch1.medtimer.database

import com.google.gson.GsonBuilder
import com.google.gson.JsonElement
import java.time.Instant

class JSONMedicineBackup(
    private val medicineRepository: MedicineRepository,
    private val reminderRepository: ReminderRepository,
    private val tagRepository: TagRepository
) : JSONBackup<FullMedicineEntity>(FullMedicineEntity::class.java) {
    override fun createBackup(databaseVersion: Int, list: List<FullMedicineEntity>): JsonElement {
        // Correct the medicines where the reminder's instructions are null in this loop
        for (fullMedicine in list) {
            fullMedicine.reminders.filter { reminder -> reminder.instructions == null }
                .forEach { reminder -> reminder.instructions = "" }
        }
        return super.createBackup(databaseVersion, list)
    }

    override fun registerTypeAdapters(builder: GsonBuilder): GsonBuilder {
        return builder
            .registerTypeAdapter(MedicineEntity::class.java, FullDeserialize<MedicineEntity?>())
            .registerTypeAdapter(TagEntity::class.java, FullDeserialize<Any?>())
            .registerTypeAdapter(ReminderEntity::class.java, FullDeserialize<ReminderEntity?>())
    }

    override fun isInvalid(item: FullMedicineEntity?): Boolean {
        return item == null
    }

    override suspend fun applyBackup(list: List<FullMedicineEntity>) {
        reminderRepository.deleteAll()
        medicineRepository.deleteAll()
        tagRepository.deleteAll()

        var sortOrder = 1.0

        for (fullMedicine in list) {
            if (fullMedicine.medicine.sortOrder == 0.0) {
                fullMedicine.medicine.sortOrder = sortOrder++
            }
            val medicineId = medicineRepository.create(fullMedicine.medicine)
            processReminders(fullMedicine, medicineId.toInt())
            processTags(fullMedicine, medicineId.toInt())
        }
    }

    private suspend fun processTags(fullMedicine: FullMedicineEntity, medicineId: Int) {
        for (tag in fullMedicine.tags) {
            val tagId = tagRepository.create(tag).toInt()
            tagRepository.addMedicineTag(medicineId, tagId)
        }
    }

    private suspend fun processReminders(fullMedicine: FullMedicineEntity, medicineId: Int) {
        val reminders: MutableList<ReminderEntity> = mutableListOf()
        for (reminder in fullMedicine.reminders) {
            reminder.medicineRelId = medicineId
            reminder.createdTimestamp = Instant.now().toEpochMilli() / 1000
            reminders.add(reminder)
        }
        reminderRepository.createAll(reminders)
    }
}
